src/main/kotlin/com/demonwav/mcdev/platform/mcp/at/completion/AtCompletionContributor.kt

```diff
@@ -0,0 +1,285 @@
+import com.demonwav.mcdev.util.fullQualifiedName
+import com.demonwav.mcdev.util.nameAndParameterTypes
+import com.demonwav.mcdev.util.qualifiedMemberReference
+import com.demonwav.mcdev.util.simpleQualifiedMemberReference
+import com.intellij.codeInsight.completion.CompletionContributor
+import com.intellij.codeInsight.completion.CompletionParameters
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.codeInsight.completion.CompletionType
+import com.intellij.openapi.module.ModuleUtilCore
+import com.intellij.patterns.PlatformPatterns.elementType
+import com.intellij.patterns.PlatformPatterns.psiElement
+import com.intellij.patterns.PsiElementPattern
+import com.intellij.psi.JavaPsiFacade
+import com.intellij.psi.PsiClass
+import com.intellij.psi.PsiElement
+import com.intellij.psi.search.GlobalSearchScope
+import com.intellij.psi.search.PsiShortNamesCache
+import com.intellij.psi.tree.IElementType
+import com.intellij.psi.util.PsiUtilCore
+import com.intellij.util.PlatformIcons
+
+class AtCompletionContributor : CompletionContributor() {
+
+    override fun invokeAutoPopup(position: PsiElement, typeChar: Char): Boolean {
+        if (typeChar == '$') {
+            return true
+        }
+        return super.invokeAutoPopup(position, typeChar)
+    }
+
+    override fun fillCompletionVariants(parameters: CompletionParameters, result: CompletionResultSet) {
+        if (parameters.completionType != CompletionType.BASIC) {
+            return
+        }
+
+        val position = parameters.position
+        if (!PsiUtilCore.findLanguageFromElement(position).isKindOf(AtLanguage)) {
+            return
+        }
+
+        val parent = position.parent
+
+        if (AFTER_KEYWORD.accepts(parent)) {
+            handleAtClassName(parent, result)
+        } else if (AFTER_CLASS_NAME.accepts(parent)) {
+            handleAtName(parent, result)
+        } else if (AFTER_NEWLINE.accepts(parent)) {
+            handleNewLine(parent, result)
+        }
+    }
+
+    private fun handleAtClassName(element: PsiElement, result: CompletionResultSet) {
+        val text = element.text.let { it.substring(0, it.length - 19) }
+        if (text.isEmpty()) {
+            return
+        }
+
+        val currentPackage = text.substringBeforeLast('.')
+        val beginning = text.substringAfterLast('.').toLowerCase()
+
+        val module = ModuleUtilCore.findModuleForPsiElement(element) ?: return
+        val scope = GlobalSearchScope.moduleWithDependenciesAndLibrariesScope(module)
+        val project = module.project
+
+        // Short name completion
+        if (!text.contains('.')) {
+            val kindResult = result.withPrefixMatcher(KindPrefixMatcher(text))
+            val cache = PsiShortNamesCache.getInstance(project)
+
+            var counter = 0
+            for (className in cache.allClassNames) {
+                if (!className.toLowerCase().startsWith(beginning)) {
```

Comment:
`startsWith` supports a `ignoreCase` parameter too, you don't need `toLowerCase` necessarily here